R/estimateParams.R

```diff
@@ -53,3 +53,366 @@ prepareData <- function(data) {
+getResults <- function(transition, res) {
+  UseMethod("getResults")
+}
+
+#' Format Results of Parameter Estimation for Exponential Transition
+#'
+#' @param transition (`ExponentialTransition`)\cr
+#'   See [exponential_transition()] for details.
+#' @inheritParams getResults
+#'
+#' @return Returns an `ExponentialTransition` object with updated parameter estimates.
+#' @export
+#'
+#' @details
+#' Constructs an `ExponentialTransition` object with parameters estimated from likelihood maximization.
+#'
+#' @examples
+#' results <- c(1.2, 1.5, 1.6)
+#' getResults(exponential_transition(), results)
+getResults.ExponentialTransition <- function(transition, res) {
+  exponential_transition(h01 = res[1], h02 = res[2], h12 = res[3])
+}
+
+#' Format Results of Parameter Estimation for Weibull Transition
+#'
+#' @param transition (`WeibullTransition`)\cr
+#'   See [weibull_transition()] for details.
+#' @inheritParams getResults
+#'
+#' @return Returns a `WeibullTransition` object with updated parameter estimates.
+#' @export
+#'
+#' @details
+#' Constructs a `WeibullTransition` object with parameters estimated from likelihood maximization.
+#'
+#' @examples
+#' results <- c(1.2, 1.5, 1.6, 2, 2.5, 1)
+#' getResults(weibull_transition(), results)
+getResults.WeibullTransition <- function(transition, res) {
+  weibull_transition(
+    h01 = res[1], h02 = res[2], h12 = res[3],
+    p01 = res[4], p02 = res[5], p12 = res[6]
+  )
+}
+
+#' Estimate Parameters of the Multistate Model Using Clinical Trial Data
+#'
+#' @param data (`data.frame`)\cr
+#'   Data frame in the format produced by [getOneClinicalTrial()].
+#' @param transition (`transitionParameters` object)\cr
+#'   Object specifying the assumed distribution of transition hazards.
+#'   Initial parameters for optimization can be specified here.
+#'   See [exponential_transition()] or [weibull_transition()] for details.
+#'
+#' @return Returns a `transitionParameters` object with the estimated parameters.
+#' @export
+#'
+#' @details
+#' This function estimates parameters for transition models using clinical trial data.
+#' The `transition` object can be initialized with starting values for parameter estimation.
+#' It uses [stats::optim()] to optimize the parameters.
+#'
+#' @examples
+#' transition <- exponential_transition(h01 = 2, h02 = 1.4, h12 = 1.6)
+#' simData <- getOneClinicalTrial(
+#'   nPat = c(30), transitionByArm = list(transition),
+#'   dropout = list(rate = 0.3, time = 12),
+#'   accrual = list(param = "time", value = 1)
+#' )
+#' # Initialize transition with desired starting values for optimization:
+#' transition <- exponential_transition(h01 = 1.2, h02 = 1.5, h12 = 1.6)
+#' estimate <- estimateParams(simData, transition)
+estimateParams <- function(data, transition) {
+  data <- prepareData(data)
+
+  res <- stats::optim(
+    par = getInit(transition),
+    fn = getTarget,
+    method = "L-BFGS-B",
+    lower = 1e-3,
+    data = data,
+    transition = transition
+  )$par
```

Comment:
so here it will look something like this:
```suggestion

  par <- getInit(transition)
  target <- getTarget(transition)
  res <- stats::optim(
    par = par,
    fn = target,
    method = "L-BFGS-B",
    lower = 1e-3,
    data = data
  )$par
```